drivers: spi: fix Ambiq apollo4p SPI configuration and transmission issues.

aaronyegx opened the pull request:
Fix#1: The SPI mode was configured improperly. The AM_HAL_IOM_SPI_MODE_3 and AM_HAL_IOM_SPI_MODE_0 cases would never be entered as expected in previous implementation. 
-- (SPI_MODE_COPL && SPI_MODE_CPHA) is always 0 so the "if (config->operation & (SPI_MODE_CPOL && SPI_MODE_CPHA))" can only be false, then AM_HAL_IOM_SPI_MODE_3 can never be configured.

Fix#2: The configured SPI shall not be configured again, otherwise it would cause SPI transmission failure.

Fix#3: The SPI instruction length can only be 0~5, it shall not be assigned to tx_len which is larger than 5 in the RX transmission.
Always use am_hal_iom_blocking_transfer() and specify clearly the TX/RX direction for SPI transmission.
Shall hold CS to continue to receive expected response data after instruction transmission.

With previous implementation:
--e.g., when the tx_buf[] = {0x01, 0x02, 0x03, 0x04, 0x05}, the actual transmitted data will be (0x0, 0x0, 0x0, 0x01, 0x02, 0x03, 0x04, 0x05) which is unexpected.
--e.g., when the tx_buf[] = {0x01, 0x02, 0x03, 0x04, 0x05, 0x06}, the SPI cannot transmit any data because instruction length is incorrectly set to tx_len 6. The instruction length can be 0~5.

With the Fix#3, the SPI can transmit and receive data as expected.

tbursztyka reviewed (commented):
please add your Signed-off-by in the commit messages and apply checkpatch's comments.

Beside that, changes seems fine

aaronyegx committed:
drivers: spi: Fix Ambiq apollo4p SPI mode configuration.

The commit fixes the SPI mode improper configuration.
Otherwise the MODE_3 and MODE_0 cases would never be entered as expected.

Signed-off-by: Aaron Ye <aye@ambiq.com>

aaronyegx committed:
drivers: spi: Fix Ambiq apollo4p SPI reconfiguration.

The configured SPI should not be configured again.
Otherwise it would cause SPI transmission failure.

Signed-off-by: Aaron Ye <aye@ambiq.com>

aaronyegx committed:
drivers: spi: Fix the Ambiq apollo4p SPI transmission failure.

The instruction length can only be 0~5.
Use am_hal_iom_blocking_transfer and specify clearly the TX/RX direction.
Hold CS to continue to RX expected response after instruction transmission.

Signed-off-by: Aaron Ye <aye@ambiq.com>

aaronyegx commented:
> please add your Signed-off-by in the commit messages and apply checkpatch's comments.
> 
> Beside that, changes seems fine

Updated the commit messages and comments in the code.

aaronyegx commented:
Hi @tbursztyka @fkokosinski, can you help review this PR? Thanks very much.

aaronyegx commented:
Hi @fkokosinski @tbursztyka @tgorochowik @teburd, could you please advise what I can do to move forward this pull request? I'm sorry I'm new to Zephyr contribution so I'm not sure whether I miss anything. The code change is really necessary to make the driver work properly.

fkokosinski reviewed (approved):
LGTM

aaronyegx commented:
Hi @tbursztyka @tgorochowik @teburd, could you take a look at this PR, please? Thanks so much.

aaronyegx commented:
> please add your Signed-off-by in the commit messages and apply checkpatch's comments.
> 
> Beside that, changes seems fine

Hi @tbursztyka , could you take another look at this PR, please? The PR did not introduce code changes since your last review, just updated the commit message and format as you requested. Thanks so much.

aaronyegx commented:
Hi @tbursztyka @tgorochowik @teburd, could you help take a look at this PR and let me know if anything needs to be updated to move forward it? Thank you very much.
